### PR TITLE
[SPARK-33949][SQL] Make approx_count_distinct result consistent whether Optimize rule exists or not

### PR DESCRIPTION
**Pull request by ulysses-you**
### What changes were proposed in this pull request?

Change `HyperLogLogPlusPlus.validateDoubleLiteral` from match to eval.

### Why are the changes needed?

This code will fail because foldable value not fold, we should keep result consistent whether Optimize rule exists or not.
```
val excludedRules = Seq(ConstantFolding, ReorderAssociativeOperator).map(_.ruleName)
withSQLConf(SQLConf.OPTIMIZER_EXCLUDED_RULES.key -> excludedRules.mkString(",")) {
  sql("select approx_count_distinct(1, 0.01 + 0.02)")
}
```

### Does this PR introduce _any_ user-facing change?

Maybe yes.

### How was this patch tested?

Add test.

### CODE REVIEW

**Comment by ulysses-you** [comment]
cc @maropu @dongjoon-hyun @HyukjinKwon

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala [hunk 1]
```diff
@@ -141,9 +141,11 @@ case class HyperLogLogPlusPlus(
 
 object HyperLogLogPlusPlus {
   def validateDoubleLiteral(exp: Expression): Double = exp match {
-    case Literal(d: Double, DoubleType) => d
-    case Literal(dec: Decimal, _) => dec.toDouble
+    case e if e.foldable && e.dataType == DoubleType =>
+      e.eval().asInstanceOf[Double]
+    case e if e.foldable && e.dataType.isInstanceOf[DecimalType] =>
+      e.eval().asInstanceOf[Decimal].toDouble
```
Comment (maropu) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala:147]: Ah, nice catch! Is this a last place having the same issue?

**Comment by maropu** [comment]
This is a bug fix, isn't it? It seems the jira has the `improvement` type though.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala (hunk 1 above)
Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala:147]: I believed it is but I just find a new at `FirstLast`.

Comment (ulysses-you) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala:147]: https://github.com/apache/spark/blob/3a299aa6480ac22501512cd0310d31a441d7dfdc/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/First.scala#L124-L129

Comment (maropu) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HyperLogLogPlusPlus.scala:147]: Ah, ok. Could you fix that, too, in a separate PR?

**Review by maropu — APPROVED** [review]

**Comment by ulysses-you** [comment]
Sorry @maropu  I just realize this is a mistake, that expression is not a child and will never be folded by Optimizer. This PR is redundant.

I checked again, `rand/randn` seems is the only one place has such issue. Sorry for the mistake, maybe I need to wash my head : (

**Pull request closed without merging**